File: Week03/LeetCode-0219.cpp
#include <algorithm>
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        /// O(NlogN) algorithm
        vector<pair<int, int>> A;
        for (int i=0; i<nums.size(); i++)
            A.push_back({nums[i], i});

        // Sort: O(NlogN)
        std::sort(A.begin(), A.end());
        
        // Walk over A
        for (int i=1; i<A.size(); i++){
            if (A[i].first == A[i-1].first)
                if (std::abs(A[i].second-A[i-1].second) <= k)
                    return true;
        } // end-for
        
        return false;
    } // end-containsNearbyDuplicate
};

File: Week03/LeetCode-0219.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        """
        O(NlogN) algorithm
        """
        A = []
        for i in range(len(nums)):
            A.append([nums[i], i])            

        # Sort A
        A.sort()
            
        # Walk over the sorted array
        for i in range(1, len(A)):
            if A[i][0] == A[i-1][0]:
                if abs(A[i][1]-A[i-1][1]) <= k:
                    return True            
        return False
        